_posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd

```diff
@@ -0,0 +1,347 @@
+scratch](https://blogs.rstudio.com/ai/posts/2020-10-01-torch-network-from-scratch),
+using nothing but `torch` *tensors*. Predictions, loss, gradients,
+weight updates -- all these things we've been computing ourselves.
+Today, we make a significant change: Namely, we spare ourselves the
+cumbersome calculation of gradients, and have `torch` do it for us.
+
+Prior to that though, let's get some background.
+
+## Automatic differentiation with *autograd*
+
+`Torch` uses a module called *autograd* to
+
+(1) record operations performed on tensors, and
+
+(2) store what will have to be done to obtain the corresponding
+    gradients, once we're entering the backward pass.
+
+These prospective actions are stored internally as functions, and when
+it's time to compute the gradients, these functions are applied in
+order: Application starts from the output node, and calculated gradients
+are successively being *propagated* *back* through the network. This is
+a form of *reverse mode automatic differentiation*.
+
+#### *Autograd* basics
+
+As users, we can see a bit of the implementation. As a prerequisite for
+this "recording" to happen, tensors have to be created with
+`requires_grad = TRUE`. E.g.,
+
+```{r}
+library(torch)
+
+x <- torch_ones(2, 2, requires_grad = TRUE)
+```
+
+To be clear, `x` now is a tensor *with respect to which* gradients have
+to be calculated -- normally, a tensor representing a weight or a bias,
+not the input data [^1]. If subsequently we perform some operation on
+that tensor, assigning the result to `y`,
+
+[^1]: Unless we *want* to change the data, as when generating
+    adversarial examples.
+
+```{r}
+y <- x$mean()
+```
+
+we find that `y` now has a non-empty `grad_fn` that tells `torch` how to
+compute the gradient of `y` with respect to `x`:
+
+```{r}
+y$grad_fn
+```
+
+    MeanBackward0
+
+Actual *computation* of gradients is triggered by calling `backward()`
+on the output tensor.
+
+```{r}
+y$backward()
+```
+
+After `backward()` has been called, `x` has a non-null field termed
+`grad` that stores the gradient of `y` with respect to `x`:
+
+```{r}
+x$grad
+```
+
+    torch_tensor 
+     0.2500  0.2500
+     0.2500  0.2500
+    [ CPUFloatType{2,2} ]
+
+With longer chains of computations, we can take a glance at how `torch`
+builds up a graph of backward operations. Here is a slightly more
+complex example -- feel free to skip if you're not the type who just
+*has* to peek into things for them to make sense.
+
+#### Digging deeper
+
+We build up a simple graph of tensors, with "inputs" `x1` and `x2` being
+connected to "output" `out` by "intermediaries" `y` and `z`.
+
+```{r}
+x1 <- torch_ones(2, 2, requires_grad = TRUE)
+x2 <- torch_tensor(1.1, requires_grad = TRUE)
+
+y <- x1 * (x2 + 2)
+
+z <- y$pow(2) * 3
+
+out <- z$mean()
+```
+
+To save memory, by default intermediate gradients are not stored.
+Calling `retain_grad()` on a tensor allows to deviate from this default;
```

Comment:
```suggestion
Calling `retain_grad()` on a tensor allows to deviate from this default.
```

Reply:
With technical writing, I think it's a good thing to simplify, so I sort of "factor out" semi-colon connectors where possible.